src/ec/curve25519/ed25519.rs

```diff
@@ -202,6 +202,100 @@ impl<'a> Ed25519KeyPair {
     }
 }
 
+/// The components of an Ed25519 key pair. This type cannot be used
+/// for signing, see [Ed25519KeyPair](./Ed25519KeyPair.html) instead.
+pub struct Ed25519KeyPairComponents {
+    /// The 32-byte private part of this pair, i.e. the seed.
+    pub private_key: [u8; 32],
+    /// The 32-byte public part of this key.
+    pub public_key: [u8; 32],
+}
+
+impl<'a> Ed25519KeyPairComponents {
+    /// Constructs an Ed25519 key pair by parsing an unencrypted PKCS#8 v2
+    /// Ed25519 private key.
+    ///
+    /// The input must be in PKCS#8 v2 format, and in particular it must contain
+    /// the public key in addition to the private key. `from_pkcs8()` will
+    /// verify that the public key and the private key are consistent with each
+    /// other.
+    ///
+    /// If you need to parse PKCS#8 v1 files (without the public key) then use
+    /// `Ed25519KeyPair::from_pkcs8_maybe_unchecked()` instead.
+    pub fn from_pkcs8(input: untrusted::Input)
+                      -> Result<Ed25519KeyPairComponents, error::Unspecified> {
+        let (seed, public_key) = unwrap_pkcs8(pkcs8::Version::V2Only, input)?;
+        Self::from_seed_and_public_key(seed, public_key.unwrap())
+    }
+
+    /// Constructs an Ed25519 key pair from the private key seed `seed` and its
+    /// public key `public_key`.
+    ///
+    /// It is recommended to use `Ed25519KeyPair::from_pkcs8()` instead.
+    ///
+    /// The private and public keys will be verified to be consistent with each
+    /// other. This helps avoid misuse of the key (e.g. accidentally swapping
+    /// the private key and public key, or using the wrong private key for the
+    /// public key). This also detects any corruption of the public or private
+    /// key.
+    pub fn from_seed_and_public_key(seed: untrusted::Input,
+                                    public_key: untrusted::Input)
+            -> Result<Ed25519KeyPairComponents, error::Unspecified> {
+        let pair = Self::from_seed_unchecked(seed)?;
+
+        // This implicitly verifies that `public_key` is the right length.
+        // XXX: This rejects ~18 keys when they are partially reduced, though
+        // those keys are virtually impossible to find.
+        if public_key != pair.public_key_bytes() {
+            return Err(error::Unspecified);
+        }
+
+        Ok(pair)
+    }
+
+    /// Constructs a Ed25519 key pair from the private key seed `seed`.
+    ///
+    /// It is recommended to use `Ed25519KeyPair::from_pkcs8()` instead. When
+    /// that is not practical, it is recommended to use
+    /// `Ed25519KeyPair::from_seed_and_public_key()` instead.
+    ///
+    /// Since the public key is not given, the public key will be computed from
+    /// the private key. It is not possible to detect misuse or corruption of
+    /// the private key since the public key isn't given as input.
+    pub fn from_seed_unchecked(seed: untrusted::Input)
```

Comment:
Are you actually using this function in your key server? I suspect you're not, if not, let's remove it. Otherwise, I am curious why you need it, as I suspect you shouldn't need it.